### PR TITLE
Add Azure SDK DI doc

### PR DESCRIPTION
**Pull request by lilyjma**
Incorporate dependency injection content from following blog post: https://devblogs.microsoft.com/azure-sdk/best-practices-for-using-azure-sdk-with-asp-net-core/

@pakrym please review this article

[Staging Link](https://review.docs.microsoft.com/en-us/dotnet/azure/dependency-injection?branch=pr-en-us-24209)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=24209) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=24209) before we can accept your contribution.<br/><br/>:x: lilyjma [sign now](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=24209)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/docs?pullRequest=24209) it.</sub>

**Review by pakrym — APPROVED** [review]
LGTM

File: docs/azure/TOC.yml [hunk 1]
```diff
@@ -3,77 +3,79 @@
...
-        href: configure-visual-studio.md
-      - name: Configure VS Code for Azure development (cross-platform)
-        href: configure-vs-code.md
-      - name: Install the Azure CLI
-        href: install-azure-cli.md
-      - name: Install additional tools
-        href: azure-tools.md
-      - name: Development setup checklist
-        href: dotnet-dev-env-checklist.md                
+      - name: Introduction to Azure and .NET
```
Comment (pakrym) [inline docs/azure/TOC.yml:6]: Do we want to reformat all these items?

Comment (lilyjma) [reply docs/azure/TOC.yml:6]: I think VSCode reformatted all the indents when I hit ctrl-s.

**Review by CamSoper — CHANGES_REQUESTED** [review]
I made suggestions to fix the spacing in a first few spots in the TOC, but I'm going to need you to correct the spacing in YAML before I can merge this. I suggest reverting that one file and trying again, perhaps with a different editor.

File: docs/azure/TOC.yml [hunk 2]
```diff
@@ -3,77 +3,79 @@
...
+            - name: Configure Visual Studio for Azure development (Windows)
+              href: configure-visual-studio.md
+            - name: Configure VS Code for Azure development (cross-platform)
+              href: configure-vs-code.md
+            - name: Install the Azure CLI
+              href: install-azure-cli.md
+            - name: Install additional tools
+              href: azure-tools.md
+            - name: Development setup checklist
+              href: dotnet-dev-env-checklist.md
```
[the remark this answers is not in the document]
Comment (lilyjma) [reply docs/azure/TOC.yml:23]: No I didn't. I hit ctrl-s in VSCode and the spacing automatically changed.

[the remark this answers is not in the document]
Comment (lilyjma) [reply docs/azure/TOC.yml]: committed your suggestions

File: docs/azure/dependency-injection.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
+---
+title: Dependency injection with the Azure SDK for .NET
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:2]:
```suggestion
title: Dependency injection with the Azure .NET SDK
```

File: docs/azure/dependency-injection.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+description: Learn how to use dependency injection with the Azure SDK for .NET client libraries
+ms.date: 05/12/2021
+ms.custom: devx-track-dotnet
+ms.author: pakrym
+author: pakrym
+---
+# Dependency injection with the Azure SDK for .NET
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:9]:
```suggestion

# Dependency injection with the Azure .NET SDK
```

File: docs/azure/dependency-injection.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
+---
+title: Dependency injection with the Azure SDK for .NET
+description: Learn how to use dependency injection with the Azure SDK for .NET client libraries
+ms.date: 05/12/2021
+ms.custom: devx-track-dotnet
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:5]:
```suggestion
```

Why is this needed? This is in the .NET docs, not the Azure repo - it's my understanding that the `devx-track-*` metadata is only for the Azure docs.

File: docs/azure/dependency-injection.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+ms.author: pakrym
+author: pakrym
+---
+# Dependency injection with the Azure SDK for .NET
+
+This article demonstrates how to register Azure service clients from the [latest .NET Azure SDKs](https://azure.github.io/azure-sdk/releases/latest/index.html) in an ASP.NET Core application. Every ASP.NET Core application starts by booting up the application using the instructions provided in the `Startup` class. This includes a `ConfigureServices` method that is an ideal place to configure clients.
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:11]:
```suggestion
This article demonstrates how to register Azure service clients from the [latest Azure .NET SDKs](https://azure.github.io/azure-sdk/releases/latest/index.html) in an ASP.NET Core application. Every ASP.NET Core application starts by booting up the application using the instructions provided in the `Startup` class. This includes a `ConfigureServices` method that is an ideal place to configure clients.
```

File: docs/azure/dependency-injection.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+## Use the registered clients
+
+With the clients registered in `Startup`, you can now use them:
+
+```csharp
+[ApiController]
+[Route("[controller]")]
+public class MyApiController : ControllerBase
+{
+  private readonly BlobServiceClient blobServiceClient;
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:64]:
```suggestion
  private readonly BlobServiceClient _blobServiceClient;
```

File: docs/azure/dependency-injection.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+  public MyApiController(BlobServiceClient blobServiceClient)
+  {
+    this.blobServiceClient = blobServiceClient;
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:68]:
```suggestion
    _blobServiceClient = blobServiceClient;
```

File: docs/azure/dependency-injection.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+  }
+
+  /// Get a list of all the blobs in the demo container
+  [HttpGet]
+  public async Task<IEnumerable<string>> Get()
+  {
+    var containerClient = this.blobServiceClient.GetBlobContainerClient("demo");
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:75]:
```suggestion
    var containerClient = _blobServiceClient.GetBlobContainerClient("demo");
```

File: docs/azure/dependency-injection.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+    var results = new List<string>();
+    await foreach (BlobItem blob in containerClient.GetBlobsAsync()) 
+    {
+      results.Add(blob.Name);
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:79]:
```suggestion
        results.Add(blob.Name);
```

File: docs/azure/dependency-injection.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    }
+    return results.ToArray();
+  }
+}
+```
+
+## Store configuration separately from code
+
+Recall in the [Register client](#register-client) section above, you explicitly specify the `keyVaultUrl` and `storageUrl`. This could cause problems when you run code against different environments during development and production. The ASP.NET Core team suggests [storing such configurations in environment dependent JSON files](/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.1#default-configuration). Thus, you can have an `appsettings.Development.json` file with one set of settings and an `appsettings.Production.json` with another set of configurations. The format of the file is:
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:88]:
```suggestion
In the [Register client](#register-client) section, you explicitly specify the `keyVaultUrl` and `storageUrl`. This could cause problems when you run code against different environments during development and production. The .NET team suggests [storing such configurations in environment dependent JSON files](/dotnet/core/extensions/configuration-providers#json-configuration-provider). Thus, you can have an _appsettings.Development.json_ file with one set of settings and an _appsettings.Production.json_ with another set of configurations. The format of the file is:
```

We should be pointing more broadly to the .NET config docs here. The Azure .NET SDK is not only used in ASP.NET Core apps, it's applicable to all .NET workloads and this could be misleading otherwise.

File: docs/azure/dependency-injection.md [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+  "KeyVault": {
+    "VaultUri": "https://mykeyvault.vault.azure.net"
+  },
+  "Storage": {
+    "ServiceUri": "https://mydemoaccount.storage.windows.net"
+  }
+}
+```
+
+You can add any options from the [ClientOptions](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/ClientOptions.cs) into the `AzureDefaults` section. One of the options is the retry policy. See the section [Configure a new try policy](#configure-a-new-retry-policy).
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:111]:
```suggestion
You can add any options from the <xref:Azure.Core.ClientOptions> into the `AzureDefaults` section. One of the options is the retry policy. For more information, see [Configure a new try policy](#configure-a-new-retry-policy).
```

Use an xref instead, which correctly points to the API ref docs. For related content links, the MS guidance is to phrase it as "For more information, see [Title to the thing](the link)."

File: docs/azure/dependency-injection.md [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+  services.AddAzureClients(builder =>
+  {
+    builder.AddBlobServiceClient(Configuration.GetSection("PublicStorage"));
+    builder.AddBlobServiceClient(Configuration.GetSection("PrivateStorage"))
+      .WithName("PrivateStorage");
+  });
+}
+```
+
+In your controllers, you can access the named service clients using the `IAzureClientFactory`:
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:153]:
```suggestion
In your controllers, you can access the named service clients using the <xref:Microsoft.Extensions.Azure.IAzureClientFactory%601?displayProperty=nameWithType>:
```

Use xref and qualify type

File: docs/azure/dependency-injection.md [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+
+```csharp
+public class HomeControllers : Controller
+{
+  private BlobServiceClient publicStorage, privateStorage;
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:158]:
```suggestion
  private readonly BlobServiceClient _publicStorage;
  private readonly BlobServiceClient _privateStorage;
```

These should be `readonly`, on their own line, and follow the proper naming conventions, leading prefix of `_` for private fields.

File: docs/azure/dependency-injection.md [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+  public HomeController(BlobServiceClient defaultClient, IAzureClientFactory<BlobServiceClient> clientFactory)
+  {
+    this.publicStorage = defaultClient;
+    this.privateStorage = clientFactory.GetClient("PrivateStorage");
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:163]:
```suggestion
      _publicStorage = defaultClient;
      _privateStorage = clientFactory.GetClient("PrivateStorage");
```

File: docs/azure/dependency-injection.md [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+  }
+}
+```
+
+The un-named service client is still available in the same way as before. Named clients are additive to this.
+
+## Configure a new retry policy
+
+At some point, you might want to change the default settings for a service client. You may want different retry settings or to use a different service API version, for example. You can set the retry settings globally or on a per service basis. Say you have the following `appsettings.json` file:
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:172]:
```suggestion
At some point, you might want to change the default settings for a service client. You may want different retry settings or to use a different service API version, for example. You can set the retry settings globally or on a per service basis. Say you have the following _appsettings.json_ file:
```

Referencing a file name should be italic, instead of code fenced - per MS style guidelines.

File: docs/azure/dependency-injection.md [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+  services.AddAzureClients(builder =>
+  {
+    builder.AddBlobServiceClient(Configuration.GetSection("PublicStorage"));
+    builder.AddBlobServiceClient(Configuration.GetSection("PrivateStorage"))
+      .WithName("PrivateStorage");
+  });
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:149]:
```suggestion
    services.AddAzureClients(builder =>
    {
        builder.AddBlobServiceClient(Configuration.GetSection("PublicStorage"));
        builder.AddBlobServiceClient(Configuration.GetSection("PrivateStorage"))
            .WithName("PrivateStorage");
    });
```
Four spaces, per style preference for .NET docs.

File: docs/azure/dependency-injection.md [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+    builder.AddBlobServiceClient(Configuration.GetSection("CustomStorage"))
+      .WithName("CustomStorage")
+      .ConfigureOptions(options => {
+        options.Retry.Mode = Azure.Core.RetryMode.Exponential;
+        options.Retry.MaxRetries = 5;
+        options.Retry.MaxDelay = TimeSpan.FromSections(120);
+      });
+  });
+}
+```
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:220]:
```suggestion
```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddAzureClients(builder =>
    {
        // Establish the global defaults
        builder.ConfigureDefaults(Configuration.GetSection("AzureDefaults"));
        builder.UseCredential(new DefaultAzureCredential());
    
        // A Key Vault Secrets client using the global defaults
        builder.AddSecretClient(Configuration.GetSection("KeyVault"));
    
        // A Storage client with a custom retry policy
        builder.AddBlobServiceClient(Configuration.GetSection("Storage"))
            .ConfigureOptions(options => options.Retry.MaxRetries = 10);
    
        // A named storage client with a different custom retry policy
        builder.AddBlobServiceClient(Configuration.GetSection("CustomStorage"))
            .WithName("CustomStorage")
            .ConfigureOptions(options =>
            {
                options.Retry.Mode = Azure.Core.RetryMode.Exponential;
                options.Retry.MaxRetries = 5;
                options.Retry.MaxDelay = TimeSpan.FromSections(120);
            });
    });
}
```
```

File: docs/azure/dependency-injection.md [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+
+You can also place policy overrides in the `appsettings.json` file:
```
Comment (IEvangelist) [inline docs/azure/dependency-injection.md:222]:
```suggestion
You can also place policy overrides in the _appsettings.json_ file:
```

**Review by IEvangelist — CHANGES_REQUESTED** [review]
This is an awesome doc, thank you @lilyjma - I have some nits, but several things that should be addressed before we merge. We should probably add a **See also** section H2 with some additional links to help guide the readers to related content, such as a link to the [Dependency injection in .NET](https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection).

File: docs/azure/dependency-injection.md (hunk 5 above)
Comment (lilyjma) [reply docs/azure/dependency-injection.md:5]: oh I added because other docs seem to have that as well. I'll remove it if it's not needed.

Comment (scottaddie) [reply docs/azure/dependency-injection.md]: Confirmed that it's not needed.

File: docs/azure/dependency-injection.md (hunk 7 above)
Comment (lilyjma) [reply docs/azure/dependency-injection.md]: fixed in the latest commit that adds **See also** section

File: docs/azure/dependency-injection.md (hunk 15 above)
Comment (lilyjma) [reply docs/azure/dependency-injection.md]: fixed in the latest commit that adds **See also** section

File: docs/azure/dependency-injection.md [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+      "maxRetries": 10
+    }
+  }
+}
+```
+
+## See also
+
+- [Dependency injection in ASP.NET Core](https://docs.microsoft.com/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-3.0)
+- [Configuration in ASP.NET Core](https://docs.microsoft.com/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.0)
```
Comment (lilyjma) [inline docs/azure/dependency-injection.md:239]: @scottaddie and @IEvangelist what do you think of linking to these articles?

File: docs/azure/dependency-injection.md [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
```
Comment (scottaddie) [inline docs/azure/dependency-injection.md:239]:
```suggestion
- [Dependency injection in .NET](/dotnet/core/extensions/dependency-injection)
- [Dependency injection in ASP.NET Core](/aspnet/core/fundamentals/dependency-injection)
- [Configuration in ASP.NET Core](/aspnet/core/fundamentals/configuration)
```

File: docs/azure/dependency-injection.md (hunk 20 above)
Comment (scottaddie) [reply docs/azure/dependency-injection.md:239]: Great idea! See my suggestion below.

Comment (IEvangelist) [reply docs/azure/dependency-injection.md:239]: I think we should be linking to the .NET specific articles here, both of which do have links to the ASP.NET Core docs if needed. This doc will be used by any workload, not just web apps. As such is should be more broad and encompassing of all .NET workloads where the Azure .NET SDK can be used.

- [Dependency injection in .NET](/dotnet/core/extensions/dependency-injection)
- [Configuration in .NET](/dotnet/core/extensions/configuration)

**Review by IEvangelist — CHANGES_REQUESTED** [review]

**Review by IEvangelist — APPROVED** [review]
Looks good to me, we'll :shipit: - thank you 🙏

**Review by CamSoper — APPROVED** [review]

File: docs/azure/dependency-injection.md (hunk 20 above)
Comment (lilyjma) [reply docs/azure/dependency-injection.md]: @IEvangelist sorry for the delay. I think that's a good idea

**Comment by lilyjma** [comment]
Thanks @IEvangelist , @CamSoper and @scottaddie  for all the help!
